File: python/rest-api/rest_api.py
import sqlite3
import json

class RestAPI(object):
    def __init__(self, database=None):
        self.db = sqlite3.connect(':memory:')
        self.db.row_factory = sqlite3.Row
        self.init_db(database)

    def add_user(self, name):
        cursor = self.db.cursor()
        cursor.execute('INSERT INTO users(name) VALUES(?)', (name,))
        self.db.commit()
        user_id = self.get_user_id(cursor, name)
        return json.dumps(self.get_user(cursor, user_id))
    
    def get(self, url, payload=None):
        if url.lower() == '/users':
            cursor = self.db.cursor()
            if payload == None:
                # basic list of everything (sorted by name)                
                cursor.execute('SELECT id FROM users ORDER BY name')
                rows = cursor.fetchall()
                users = []
                for row in rows:
                    users.append(self.get_user(cursor, row['id']))
                return json.dumps({'users': users})
            else:
                # get a list of users, sorted by name but filtered to just
                # the user list passed in.
                parameters = json.loads(payload)
                if 'users' in parameters:
                    return self.get_user_list(parameters['users'])
                else:
                    raise Exception('Users not found in payload')
                    return None
        else:
            raise Exception('Not a recognized get request')
            return None
        return None 

    def get_user_list(self, names):
        users = []
        cursor = self.db.cursor()        
        for name in sorted(names):
            user_id = self.get_user_id(cursor, name)
            users.append(self.get_user(cursor, user_id))
        return json.dumps({'users': users})           

    def post(self, url, payload=None):
        if url.lower() == '/add':
            if payload == None:
                raise Exception('Add called without a name')
            else:
                parameters = json.loads(payload)
                if 'user' in parameters:
                    return self.add_user(parameters['user'])
                else:
                    return None
        elif url.lower() == '/iou':
            if payload == None:
                raise Exception('IOU called without parameters')
            else:
                parameters = json.loads(payload)
                if 'lender' in parameters:
                    lender = parameters['lender']
                else:
                    raise Exception('IOU called without a lender.')
                
                if 'borrower' in parameters:
                    borrower = parameters['borrower']
                else:
                    raise Exception('IOU called without a borrower.')

                if 'amount' in parameters:
                    amount = parameters['amount']
                else:
                    raise Exception('IOU called without an amount.')
                
                cursor = self.db.cursor()
                borrower_id = self.get_user_id(cursor, borrower)
                lender_id = self.get_user_id(cursor, lender)
                cursor.execute('INSERT INTO loans(lender_id, borrower_id, amount) VALUES(?, ?, ?)', (lender_id, borrower_id, amount))
                self.db.commit()
                users = [lender, borrower]
                return self.get_user_list(users)
        else:
            raise Exception('Not a supported post command.')
        
        return None

    def init_db(self, data=None):
        cursor = self.db.cursor()
        cursor.execute('''
        CREATE TABLE users(id INTEGER PRIMARY KEY, name TEXT)
        ''')
        self.db.commit()

        cursor.execute('''
        CREATE TABLE loans(lender_id INTEGER, borrower_id INTEGER, amount NUMERIC)
        ''')
        self.db.commit()

        if 'users' in data:
            # Add user records
            for user in data['users']:
                if 'name' in user:
                    cursor.execute('INSERT INTO users(name) VALUES(?)', (user['name'],))
            self.db.commit()

            # Add in loan records
            for user in data['users']:
                if 'name' in user and 'owes' in user:
                    # owed by is redundant, only need to insert for
                    # one or the other.
                    borrower = user['name']                    
                    borrower_id = self.get_user_id(cursor, borrower)
                    for lender, amount in user['owes'].items():
                        lender_id = self.get_user_id(cursor, lender)
                        cursor.execute('''
                        INSERT INTO loans(lender_id, borrower_id, amount) 
                        VALUES(?, ? , ?)
                        ''', (lender_id, borrower_id, amount))
                        
            self.db.commit()
    
    def get_user(self, cursor, id):
        user = {}
        name = ''
        cursor.execute('SELECT name FROM users WHERE id = ?', (id,))
        row = cursor.fetchone()
        if row == None:
            raise Exception(f'User with id = {id} not found in database.')
        else:
            name = row['name']
        user['name'] = name
        
        balance = 0
        
        # read the owes and owed by section
        cursor.execute('''
        SELECT 
            users.name AS name
            , sum(borrow.amount) AS borrow
            , sum(lend.amount) AS lend
        FROM
            users
            LEFT OUTER JOIN loans borrow ON borrow.lender_id = users.id AND borrow.borrower_id = ?
            LEFT OUTER JOIN loans lend ON lend.borrower_id = users.id AND lend.lender_id = ?
        WHERE
            borrow.borrower_id IS NOT NULL
            OR lend.lender_id IS NOT NULL
        GROUP BY
            users.name
        ORDER BY
            users.name
        ''', (id, id))
        rows = cursor.fetchall()
        owes = {}
        owed_by = {}
        for row in rows:
            if row['lend'] != None:
                lend = row['lend']
            else:
                lend = 0

            if row['borrow'] != None:
                borrow = row['borrow']
            else:
                borrow = 0

            if lend - borrow != 0:
                if lend > borrow:
                    owed_by[row['name']] = lend - borrow
                else:
                    owes[row['name']] = borrow - lend
                balance += (lend - borrow)
        
        user['owes'] = owes
        user['owed_by'] = owed_by

        user['balance'] = balance

        return user

    def get_user_id(self, cursor, name):
        id = 0
        cursor.execute('SELECT id FROM users WHERE name = ?', (name,))
        row = cursor.fetchone()
        if row == None:
            raise Exception(f'User: {name} not found in database.')
        else:
            id = row['id']
        return id

File: csharp/rest-api/RestApi.cs
using System.Text.Json;

// Use of SortedDictionary is just so the tests won't fail. 

public class RestApi
{
    class User
    {
        public string name { get; set; }
        public SortedDictionary<string, decimal> owes { get; set; }
        public SortedDictionary<string, decimal> owed_by { get; set; }
        public Decimal balance
        {
            get
            {
                return (from pair in owed_by select pair.Value).Sum() -
                    (from pair in owes select pair.Value).Sum();
            }
        }

        public User()
        {
            name = "";
            owes = new SortedDictionary<string, decimal>();
            owed_by = new SortedDictionary<string, decimal>();
        }
    }

    class Iou
    {
        public string lender { get; set; }
        public string borrower { get; set; }
        public Decimal amount { get; set; }

        public Iou()
        {
            lender = "";
            borrower = "";
            amount = 0;
        }
    }

    private List<User> database;
    public RestApi(string database)
    {
        List<User>? output = JsonSerializer.Deserialize<List<User>>(database);
        this.database = output == null ? new List<User>() : output;
    }

    public string Get(string url, string? payload = null)
    {
        switch (url)
        {
            case "/users":
                {
                    Dictionary<string, List<string>>? users = JsonSerializer.Deserialize<Dictionary<string, List<string>>?>(payload == null ? "{\"users\": []}" : payload);

                    if (users == null)
                    {
                        // All users if no list.
                        return JsonSerializer.Serialize<List<User>>(database);
                    }
                    else
                    {
                        // Just the ones in the list.
                        return JsonSerializer.Serialize<List<User>>((
                            from User user in database
                            where users["users"].Contains(user.name)
                            orderby user.name
                            select user).ToList<User>());
                    }
                }
        }
        return "";
    }

    public string Post(string url, string payload)
    {
        switch (url)
        {
            case "/add":
                {
                    Dictionary<string, string>? user = JsonSerializer.Deserialize<Dictionary<string, string>?>(payload == null ? "{\"users\": \"\"}" : payload);

                    if (user == null)
                    {
                        // No users if no list.
                        return "";
                    }
                    else
                    {
                        User? newUser = (from User u in database where u.name == user["user"] select u).FirstOrDefault<User?>();
                        if (newUser == null)
                        {
                            newUser = new User();
                            newUser.name = user["user"];
                            database.Add(newUser);
                        }
                        return JsonSerializer.Serialize<User>(newUser);
                    }
                }
            case "/iou":
                {
                    Iou? iou = JsonSerializer.Deserialize<Iou?>(payload == null ? "{}" : payload);
                    if (iou == null)
                        return "[]";

                    User? borrower = (from User user in database where user.name == iou.borrower select user).FirstOrDefault<User?>();
                    if (borrower == null)
                    {
                        borrower = new User();
                        borrower.name = iou.borrower;
                        database.Add(borrower);
                    }

                    User? lender = (from User user in database where user.name == iou.lender select user).FirstOrDefault<User?>();
                    if (lender == null)
                    {
                        lender = new User();
                        lender.name = iou.lender;
                        database.Add(lender);
                    }

                    if (!borrower.owes.ContainsKey(lender.name))
                    {
                        borrower.owes.Add(lender.name, iou.amount);
                    }
                    else
                    {
                        borrower.owes[lender.name] += iou.amount;
                    }

                    if (!lender.owed_by.ContainsKey(borrower.name))
                    {
                        lender.owed_by.Add(borrower.name, iou.amount);
                    }
                    else
                    {
                        lender.owed_by[borrower.name] -= iou.amount;
                    }

                    SimplifyDatabase();

                    // Just the ones in the list.
                    List<User> parties = (from User u in database
                                          where u.name == lender.name || u.name == borrower.name
                                          orderby u.name
                                          select u).ToList<User>();
                    return JsonSerializer.Serialize<List<User>>(parties);
                }
        }
        return "";
    }

    void SimplifyDatabase()
    {
        foreach (User user in database)
        {
            foreach (KeyValuePair<string, decimal> pair in user.owes.ToList())
            {
                if (user.owed_by.ContainsKey(pair.Key))
                {
                    // Owes and owed by.
                    Decimal delta = pair.Value - user.owed_by[pair.Key];
                    if (delta == 0)
                    {
                        // Remove the keys they are even.
                        user.owes.Remove(pair.Key);
                        user.owed_by.Remove(pair.Key);
                    }
                    else if (delta < 0)
                    {
                        // User owes remove, owedby = -delta.
                        user.owes.Remove(pair.Key);
                        user.owed_by[pair.Key] = -1 * delta;
                    }
                    else // delta > 0
                    {
                        // User owes delta, owedby removed
                        user.owes[pair.Key] = delta;
                        user.owed_by.Remove(pair.Key);
                    }
                }
            }
        }
    }
}
